### ISSUE TITLE
CI DAG

### ISSUE THREAD

**Issue by danking** [hail-is/hail#5193]
# People

@cseed @jigold @akotlar @danking @tpoterba

# Plan

We want to describe build processes like this:

```
- type: namespace
  name: ns
- type: image
  name: hail-pr-builder
  context: ../
  dockerfile: ../Dockerfile
- type: exec
  name: build-jar
  image: hail-pr-builder
  namespace: ns
  command: ["./gradlew", "test", "shadowJar"]
  outputs:
    - "build/libs/hail-all-spark.jar"
- type: exec
  name: pytests
  image: hail-pr-builder
  dependsOn:
    - build-jar
  command: ["./run-python-tests-using-input-jar.sh"]
```

A series of steps that get us there:

 - [x] (https://github.com/hail-is/hail/pull/5231) jobs may only depend on other jobs in the same batch 
 - [x] (https://github.com/hail-is/hail/pull/5232) a batch can be `closed` indicating no more jobs will be added
 - [ ] (https://github.com/hail-is/hail/pull/5233) a batch is automatically closed after 30 minutes
 - [ ] add "outputs" to batch jobs. When a batch job is complete, the batch system copies its outputs to some durable storage (e.g. GCS), this is not user-visible
 - [ ] add input/output dependencies to batch jobs. A batch job always receives the output of its parents. The data is copied (somehow) from the durable storage to the pod's filesystem at the path `/input/PARENT_JOB_NAME` and set to permissions 777. The copying is done in a secure way. In particular, the job container can be completely unprivileged (e.g. no credentials in the job's image, no credentials in the mounted volumes, no way to escalate to these credentials)
 - [ ] parse and run a series of `exec` commands that execute in parallel but no namespace dependencies and no image dependencies. This immediately puts us in a better place wrt logging. All artifacts, such as HTML reports are served through the batch outputs mechanism described above. Job file dependencies are handled exactly as described in input/output dependencies above.
 - [ ] allow "finalizer" jobs. A finalizer job executes when its parents are all complete or cancelled. It is not cancelled when its parents are cancelled.
 - [ ] add namespace dependencies. CI allocates anonymous namespaces as requested by the build process. All `exec`s are, by default, run in an anonymous namespace. CI adds a finalizer job that deletes namespaces when all relevant `exec`s are finished
 - [ ] add image dependencies. CI can create a batch job that builds a docker image with an anonymous name and pushes it to the project's GCR. CI adds a finalizer job that deletes the image when all relevant `exec`s are finished.
 - [ ] batch and notebook are parameterized by their worker namespace so they can use the namespaces described above
 - [ ] hail's build steps are parameterized in a way that permits them to use a jar not built locally on this machine (hopefully the Make PR makes this easy, otherwise we have to fool gradle into not rebuilding the jar)

To reliably handle clean up, we *must* persist batch jobs, so I think that should be either higher priority or at least happening in parallel to the above (i.e. two developers working in parallel).

 - [ ] persist batch jobs in a durable store with all of the fields in the beginning of `Job.__init__`. When batch starts up, before serving any requests, it restores its state from the durable store and then refreshes from k8s. The k8s label `hail.is/batch-instance` is retired. Instead, pods have `hail.is/batch-version` which is a monotonically increasing natural number. It is only incremented if batch is backwards incompatible with the pod specs. Probably batch should destroy any pods that are alive from an out-of-date version of batch.
 - [ ] persist CI information in a durable store [this needs more detail]
 - [ ] How do we configure the database? How do we create new tables? Should this be done in  the applications themselves or during deployment?

**Comment by akotlar** [comment]
Yes, so I would be happy to help you in any way you need on batch / persistence. Obviously the Go solution is out, but I have some thoughts, early code for a python version. We should probably just stage it, so that the first series of commits gets durable store up, and then we move to some async (or if its much easier gunicorn / falcon option, which would involve of course fewer changes).

let me know how you'd like to proceed, and where you'd like me to contribute.

**Comment by danking** [comment]
I believe the next step should be the minimal changes to move the global variables into `server/globals.py`. We also need to think about sql configuration. For now, I suppose we can just create a table manually, but we need a longer term strategy for this.

**Comment by danking** [comment]
# Notes On Container Security

Every pod has a [`PodSecurityContext`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#podsecuritycontext-v1-core) which lets us restrict linux permissions, the linux user id, group id, SELinux policies, and permitted sysctl resources. I suspect for CI purposes non-root user is fine. k8s [already limits](https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/) pods to the "safe" sysctl resources.

Each container in a pod as a [`SecurityContext`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#securitycontext-v1-core). By default containers have `privileged` set to false, meaning they don't have root-like abilities on the host system.

@cseed I don't see anyway for a pod to modify the mounted volumes, *particularly* if has no privileges to access the k8s api. If the volumes it is already given have no credentials that give access to k8s, I don't see how it could possibly ask k8s to mount a volume for it.

**Comment by cseed** [comment]
I like this plan overall.  Some comments to consider:

 - I don't think batch should persist job intermediates which I think more as a pipe (|) or a temporary file in a traditional script.  This means the downstream batch clients (CI, Pipeline) should be copying files they want to persist to something like gs.  One thing we should be sure here is that the solution doesn't involve excessive copies, e.g. we don't want to generate batch steps that just copy from a temporary persisted location in gs to a permanent location in gs.
 - We need a way to refer to individual inputs/outputs.  A bioinformatics command might output a massive data file and a report file, and we want to run a command to process or format the report file, but we don't want to copy the (unused) massive data file unnecessarily.  Copying everything is a fine start.
 - This might be covered by "parse and exec series of commands", but we want to be able to specify a series of stacked containers to execute, e.g. in the case of Pipeline, a user command to execute followed by a pipeline-controlled container with Pipeline gs credentials to copy the output to Pipeline (or user) controlled bucket.  Same for CI.
 - In the long run, we're going to want to be able to control the size of the local disk for user jobs (e.g. a bioinformatics command that needs 1TB of scratch space) separate from the host node's local file system.  This will go into the job configuration.  We probably don't need this for CI.
 - We should separate the database management from the job execution.  Probably easiest to make this a monolithic service, but we could separate them.  One will interact with batch users to update the database and watch the database, and the other to reconcile the database and k8s by running jobs and updating results.

**Comment by danking** [comment]
@cseed, digesting from this and our in-person conversation:

I added a steps about job dependencies and batch closure.

Regarding "a way to refer to individual inputs/outputs", in the original post, I gave the example of:

```
- type: exec
  name: build-jar
  image: hail-pr-builder
  namespace: ns
  command: ["./gradlew", "test", "shadowJar"]
  outputs:
    - "build/libs/hail-all-spark.jar"
- type: exec
  name: pytests
  image: hail-pr-builder
  dependsOn:
    - build-jar
  command: ["./run-python-tests-using-input-jar.sh"]
```

The outputs of a job are stated explicitly by the definition. For the `pytests` job in the example above, the input from `build-jar` is placed at `/inputs/build-jar`.


Regarding "specify a series of stacked containers to execute", I don't see a straightforward way to implement this. It's tricky enough to have a "anti-init"/"finalizer" container. Inter-job I/O will be handled by batch. The user controls the image and the command and the environment variables of the build step, so they can arrange for permission to copy results to a bucket they own. Are we worried about the setting of user's wanting to run untrustworthy software? They already run arbitrary software on cloud instances that have plenty of latent credentials. I think we can at least punt on this until other functionality is in.

Local disk sounds like a nice thing to add eventually.

Agreed, that sounds like a nice model. I'll consider it as I envision a persistent batch system.
